File: node/src/core/config-schema.ts
export type RiskLevel = 'minimal' | 'moderate' | 'aggressive';
export type CommandPolicyMode = 'allow-all' | 'approve-dangerous' | 'deny-list';
export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface ScanCustomPattern {
  name: string;
  regex: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
}

export interface ScanIgnoreRule {
  paths: string[];
  rules?: string[];
  reason?: string;
}

export interface RafterConfig {
  version: string;
  initialized: string;

  // Backend config (existing)
  backend?: {
    apiKey?: string;
    endpoint?: string;
  };

  // Agent config (new)
  agent?: {
    riskLevel: RiskLevel;
    environments: {
      openclaw?: {
        enabled: boolean;
        skillPath: string;
      };
      claudeCode?: {
        enabled: boolean;
        mcpPath: string;
      };
      codex?: {
        enabled: boolean;
        skillsDir: string;
      };
      gemini?: {
        enabled: boolean;
        configPath: string;
      };
      aider?: {
        enabled: boolean;
        configPath: string;
      };
      cursor?: {
        enabled: boolean;
        mcpPath: string;
      };
      windsurf?: {
        enabled: boolean;
        mcpPath: string;
      };
      continueDev?: {
        enabled: boolean;
        configPath: string;
      };
    };
    skills: {
      autoUpdate: boolean;
      installOnInit: boolean;
      backupBeforeUpdate: boolean;
      installedVersion?: string;
      lastChecked?: string;
    };
    commandPolicy: {
      mode: CommandPolicyMode;
      blockedPatterns: string[];
      requireApproval: string[];
      /**
       * Opt out of the project-policy floor (sable-nz4y).
       *
       * Read ONLY from the machine owner's global config — never from a
       * project `.rafter.yml`, or a repo could grant itself the permission and
       * the floor would be no floor at all. Default (absent/false) keeps the
       * floor: a project policy may tighten command policy, never loosen it.
       */
      allowProjectOverride?: boolean;

      /**
       * Positive allowlist: unanchored regexes that force a command to `low`
       * and skip the approval prompt. For the known-safe command that would
       * otherwise trip a broad risk tier -- the motivating case being
       * `git push --force-with-lease` to a feature branch on a repo whose
       * main is protected server-side.
       *
       * Three properties make this safe to put on a guard rail, and all three
       * are enforced in CommandInterceptor, not here:
       *   1. blockedPatterns always wins. An allowlist never re-opens what a
       *      deny rule closed.
       *   2. A `critical` command is never allowlistable.
       *   3. A match does not apply when the command contains a chain
       *      operator, so "^git push" cannot wave through
       *      `rm -rf / && git push`.
       */
      allowedPatterns?: string[];
    };
    outputFiltering: {
      redactSecrets: boolean;
      blockPatterns: boolean;
    };
    audit: {
      logAllActions: boolean;
      retentionDays: number;
      logLevel: LogLevel;
      logPath?: string;
    };
    notifications?: {
      webhook?: string;
      minRiskLevel?: 'high' | 'critical';
    };
    /**
     * Runtime enable/disable for the PreToolUse / pre-commit hook. Distinct from
     * `components["<platform>.hooks"]` (which tracks whether a hook is *installed*
     * in a platform's settings) — this gates whether an installed hook actually
     * acts. Default (undefined) = enabled.
     *
     * SECURITY: by design this is honored ONLY from the global
     * `~/.rafter/config.json` (machine-owner-owned) and the `RAFTER_DISABLE_*`
     * env vars — NEVER from project-local `.rafter.yml`, so a hostile repo can't
     * ship a config that silently disables a victim's hook (see hook-control.ts).
     * That is why this field lives on RafterConfig but NOT on PolicyFile.
     */
    hooks?: {
      /** Master switch. false = the hook allows everything (no scan, no command policy). */
      enabled?: boolean;
      /** Disable only the secret scan on Write/Edit/staged content; keep command policy. */
      secretScan?: boolean;
      /** Disable only command-risk interception on Bash; keep secret scanning. */
      commandPolicy?: boolean;
    };
    scan?: {
      excludePaths?: string[];
      customPatterns?: ScanCustomPattern[];
      ignore?: ScanIgnoreRule[];
      /**
       * sable-o4k — auto-update a stale rafter-managed betterleaks binary at
       * scan time. Default true. Set false (YAML: `scan.auto_update_betterleaks`)
       * to opt out, e.g. in CI that provisions its own binary.
       */
      autoUpdateBetterleaks?: boolean;
      /**
       * sable-9ddf — require explicit confirmation before a paid Plus scan
       * (`rafter run --mode plus`). Default false (undefined) — existing
       * behavior is unchanged unless opted in. When true, a Plus scan refuses
       * in a non-interactive/agent context unless `--yes` or `RAFTER_CONFIRM=1`
       * is present, and prompts when a TTY is attached.
       *
       * SECURITY: honored additively (OR) across global config and project
       * `.rafter.yml` — a project policy can turn this ON but can NEVER turn OFF
       * a gate the machine owner enabled globally. See runRemoteScan.
       * YAML key: `scan.plus_requires_approval`.
       */
      plusRequiresApproval?: boolean;
    };
    /**
     * Fine-grained per-component install state. Keys are component IDs like
     * "claude-code.hooks" or "cursor.mcp". Set by `rafter agent enable/disable`
     * and used by `rafter agent list` to distinguish explicitly-disabled
     * components from ones that were never installed.
     */
    components?: Record<string, {
      enabled: boolean;
      updatedAt?: string;
    }>;
  };
}

File: python/rafter_cli/core/config_schema.py
"""Configuration schema and defaults."""
from __future__ import annotations

import os
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Literal


# Type aliases
RiskLevel = Literal["minimal", "moderate", "aggressive"]
CommandPolicyMode = Literal["allow-all", "approve-dangerous", "deny-list"]
LogLevel = Literal["debug", "info", "warn", "error"]

CONFIG_VERSION = "1.0.0"


def _default_blocked_patterns() -> list[str]:
    from .risk_rules import DEFAULT_BLOCKED_PATTERNS
    return DEFAULT_BLOCKED_PATTERNS


def _default_require_approval() -> list[str]:
    from .risk_rules import DEFAULT_REQUIRE_APPROVAL
    return DEFAULT_REQUIRE_APPROVAL


@dataclass
class ScanCustomPattern:
    name: str
    regex: str
    severity: Literal["low", "medium", "high", "critical"] = "high"


@dataclass
class CommandPolicyConfig:
    mode: CommandPolicyMode = "approve-dangerous"
    blocked_patterns: list[str] = field(
        default_factory=lambda: list(_default_blocked_patterns())
    )
    require_approval: list[str] = field(
        default_factory=lambda: list(_default_require_approval())
    )


@dataclass
class AuditConfig:
    log_all_actions: bool = True
    retention_days: int = 30
    log_level: LogLevel = "info"


@dataclass
class NotificationsConfig:
    webhook: str | None = None
    min_risk_level: Literal["high", "critical"] = "high"


@dataclass
class ScanConfig:
    exclude_paths: list[str] = field(default_factory=list)
    custom_patterns: list[ScanCustomPattern] = field(default_factory=list)


@dataclass
class OutputFilteringConfig:
    redact_secrets: bool = True
    block_patterns: bool = True


@dataclass
class EnvironmentConfig:
    enabled: bool = False
    path: str = ""


@dataclass
class EnvironmentsConfig:
    openclaw: EnvironmentConfig = field(default_factory=lambda: EnvironmentConfig(
        path=os.path.join(os.path.expanduser("~"), ".openclaw", "skills", "rafter-security.md"),
    ))
    claude_code: EnvironmentConfig = field(default_factory=lambda: EnvironmentConfig(
        path=os.path.join(os.path.expanduser("~"), ".claude", "mcp", "rafter-security.json"),
    ))


@dataclass
class AgentConfig:
    risk_level: RiskLevel = "moderate"
    environments: EnvironmentsConfig = field(default_factory=EnvironmentsConfig)
    command_policy: CommandPolicyConfig = field(default_factory=CommandPolicyConfig)
    output_filtering: OutputFilteringConfig = field(default_factory=OutputFilteringConfig)
    audit: AuditConfig = field(default_factory=AuditConfig)
    notifications: NotificationsConfig = field(default_factory=NotificationsConfig)
    scan: ScanConfig = field(default_factory=ScanConfig)


@dataclass
class BackendConfig:
    api_key: str | None = None
    endpoint: str = "https://rafter.so/api/"


@dataclass
class RafterConfig:
    version: str = CONFIG_VERSION
    initialized: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    backend: BackendConfig = field(default_factory=BackendConfig)
    agent: AgentConfig = field(default_factory=AgentConfig)


def get_default_config() -> RafterConfig:
    return RafterConfig()


def get_rafter_dir() -> Path:
    return Path.home() / ".rafter"


def get_config_path() -> Path:
    return get_rafter_dir() / "config.json"


def get_audit_log_path() -> Path:
    return get_rafter_dir() / "audit.jsonl"
